### ISSUE TITLE
[Bug]: aws_ecr_repository panic when repo policy prevents read

### ISSUE THREAD

**Issue by Clete2** [hashicorp/terraform-provider-aws#29989]
### Terraform Core Version

1.4.0

### AWS Provider Version

4.58.0

### Affected Resource(s)

* aws_ecr_repository

### Expected Behavior

An error should be returned by the provider that the repository's policy prevents the read of the resource

### Actual Behavior

The AWS provider panics

### Relevant Error/Panic Output Snippet

```shell
~> terraform apply
data.aws_caller_identity.current: Reading...
aws_ecr_repository.repo: Refreshing state... [id=testing-tf-bug]
╷
│ Error: Plugin did not respond
│ 
│   with aws_ecr_repository.repo,
│   on example.tf line 1, in resource "aws_ecr_repository" "repo":
│    1: resource "aws_ecr_repository" "repo" {
│ 
│ The plugin encountered an error, and failed to respond to the plugin.(*GRPCProvider).ReadResource call. The plugin logs may contain more details.
╵
╷
│ Error: Request cancelled
│ 
│   with data.aws_caller_identity.current,
│   on example.tf line 31, in data "aws_caller_identity" "current":
│   31: data "aws_caller_identity" "current" {}
│ 
│ The plugin.(*GRPCProvider).ReadDataSource request was cancelled.
╵

Stack trace from the terraform-provider-aws_v4.58.0_x5 plugin:

panic: interface conversion: interface {} is nil, not *ecr.Repository

goroutine 42 [running]:
github.com/hashicorp/terraform-provider-aws/internal/service/ecr.resourceRepositoryRead({0xf286040?, 0xc00373ecc0}, 0xc0041cfc80, {0xddba480?, 0xc0003c3400})
        github.com/hashicorp/terraform-provider-aws/internal/service/ecr/repository.go:186 +0x12cc
github.com/hashicorp/terraform-provider-aws/internal/provider.wrappedReadContextFunc.func1({0xf286040?, 0xc00373ecc0?}, 0x0?, {0xddba480?, 0xc0003c3400?})
        github.com/hashicorp/terraform-provider-aws/internal/provider/provider.go:810 +0x3e
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).read(0xf286040?, {0xf286040?, 0xc00373ecc0?}, 0xd?, {0xddba480?, 0xc0003c3400?})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.25.0/helper/schema/resource.go:719 +0x87
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).RefreshWithoutUpgrade(0xc000ea0540, {0xf286040, 0xc00373ecc0}, 0xc0002536c0, {0xddba480, 0xc0003c3400})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.25.0/helper/schema/resource.go:1015 +0x585
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ReadResource(0xc000204168, {0xf286040?, 0xc00373eba0?}, 0xc003849280)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.25.0/helper/schema/grpc_provider.go:613 +0x4a5
github.com/hashicorp/terraform-plugin-mux/tf5muxserver.muxServer.ReadResource({0xc003163fb0, 0xc003196030, {0xc004a712a0, 0x2, 0x2}, {0x0, 0x0, 0x0}, {0x0, 0x0, ...}, ...}, ...)
        github.com/hashicorp/terraform-plugin-mux@v0.9.0/tf5muxserver/mux_server_ReadResource.go:26 +0x102
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ReadResource(0xc0021d9900, {0xf286040?, 0xc003729b30?}, 0xc002d4ea20)
        github.com/hashicorp/terraform-plugin-go@v0.14.3/tfprotov5/tf5server/server.go:748 +0x4b1
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ReadResource_Handler({0xdacb8c0?, 0xc0021d9900}, {0xf286040, 0xc003729b30}, 0xc0006f0c40, 0x0)
        github.com/hashicorp/terraform-plugin-go@v0.14.3/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:349 +0x170
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0043b52c0, {0xf295120, 0xc0005da680}, 0xc0038b5440, 0xc004b55230, 0x15a01510, 0x0)
        google.golang.org/grpc@v1.53.0/server.go:1336 +0xd23
google.golang.org/grpc.(*Server).handleStream(0xc0043b52c0, {0xf295120, 0xc0005da680}, 0xc0038b5440, 0x0)
        google.golang.org/grpc@v1.53.0/server.go:1704 +0xa2f
google.golang.org/grpc.(*Server).serveStreams.func1.2()
        google.golang.org/grpc@v1.53.0/server.go:965 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
        google.golang.org/grpc@v1.53.0/server.go:963 +0x28a

Error: The terraform-provider-aws_v4.58.0_x5 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
```

### Terraform Configuration Files

```
resource "aws_ecr_repository" "repo" {
  name = "testing-tf-bug"
}

resource "aws_ecr_repository_policy" "repo_private" {
  repository = aws_ecr_repository.repo.name
  policy     = data.aws_iam_policy_document.private_ecr.json
}

data "aws_iam_policy_document" "private_ecr" {
  statement {
    sid         = "NobodyExceptTheseRolesCanTouchThisRepo"
    effect      = "Deny"
    not_actions = ["ecr:GetRepositoryPolicy"]

    principals {
      type        = "*"
      identifiers = ["*"]
    }

    condition {
      test     = "StringNotLike"
      variable = "aws:PrincipalArn"
      values = toset([
        data.aws_iam_session_context.current.issuer_arn
      ])
    }
  }
}

data "aws_caller_identity" "current" {}

data "aws_iam_session_context" "current" {
  arn = data.aws_caller_identity.current.arn
}
```

### Steps to Reproduce

1. Login as any role which has ECR permissions
2. Apply the terraform
3. Login as any other role (will not have access to this repository due to the repo policy provided above)
4. Try to apply
5. Panic :(

### Important Factoids

I tested the above example and it reproduced the error.

### Would you like to implement a fix?

None

**Comment by Clete2** [comment]
FYI - I worked around the issue by simply going back to the first role that I assumed, adding in the second role into the policy, and then moving to the second role. IMO this issue is not urgent because the workaround should be easy to implement. The major issue is that the panic doesn't tell you what happened or why it failed.

**Comment by Clete2** [comment]
Looking at release notes, it appeared that v1.4.1 may fix this issue, but I just re-tested and the issue persists in v1.4.1 of Terraform.

**Comment by jar-b** [comment]
Fixed with 7228b595dded02a6326265150bff5c51a725dba6 (#30067).
